Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp

```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+        lastModificationTime = idbStorageModificationTime > lastModificationTime ? idbStorageModificationTime : lastModificationTime;
+    }
+
+    if (m_unifiedOriginStorageLevel <= UnifiedOriginStorageLevel::Basic) {
+        auto cacheStoragePath = CacheStorageManager::cacheStorageOriginDirectory(m_customCacheStoragePath, origin);
+        auto cacheStorageModificationTime = valueOrDefault(FileSystem::fileModificationTime(cacheStoragePath));
+        lastModificationTime = cacheStorageModificationTime > lastModificationTime ? cacheStorageModificationTime : lastModificationTime;
+    }
+
+    if (m_unifiedOriginStorageLevel <= UnifiedOriginStorageLevel::Standard) {
+        auto originFile = manager.path();
+        auto originFileModificationTime = valueOrDefault(FileSystem::fileModificationTime(originFile));
+        lastModificationTime = originFileModificationTime > lastModificationTime ? originFileModificationTime : lastModificationTime;
+    }
+
+    return lastModificationTime;
+}
+
+void NetworkStorageManager::didGetDomainsWithLastAccessedTimeForEviction(std::optional<HashMap<WebCore::RegistrableDomain, WallTime>> domainsWithLastAccessedTime)
+{
+    assertIsCurrent(workQueue());
+
+    HashMap<WebCore::SecurityOriginData, AccessRecord> originRecords;
+    uint64_t totalUsage = 0;
+    for (auto& origin : getAllOrigins()) {
+        auto usage = originStorageManager(origin, ShouldWriteOriginFile::No).quotaManager().usage();
+        totalUsage += usage;
+        WallTime accessTime;
+        if (domainsWithLastAccessedTime)
+            accessTime = domainsWithLastAccessedTime->get(WebCore::RegistrableDomain { origin.topOrigin });
+        else
+            accessTime = lastModificationTimeForOrigin(origin, originStorageManager(origin, ShouldWriteOriginFile::No));
+
+        auto& record = originRecords.ensure(origin.topOrigin, [&] {
+            return AccessRecord { };
+        }).iterator->value;
+        record.usage += usage;
+        if (record.lastAccessTime < accessTime)
+            record.lastAccessTime = accessTime;
+
+        record.clientOrigins.append(origin.clientOrigin);
+        removeOriginStorageManagerIfPossible(origin);
+    }
+
+    m_totalUsage = totalUsage;
+    performEviction(WTFMove(originRecords));
+}
+
+void NetworkStorageManager::performEviction(HashMap<WebCore::SecurityOriginData, AccessRecord>&& originRecords)
+{
+    assertIsCurrent(workQueue());
+
+    m_isEvictionScheduled = false;
+    if (!m_totalQuota || !m_totalUsage || *m_totalUsage <= *m_totalQuota)
+        return;
+
+    Vector<WebCore::SecurityOriginData> sortedOrigins;
+    for (auto& origin : originRecords.keys())
+        sortedOrigins.append(origin);
+
+    std::sort(sortedOrigins.begin(), sortedOrigins.end(), [&](const auto& a, const auto& b) {
+        return originRecords.get(a).lastAccessTime > originRecords.get(b).lastAccessTime;
+    });
```

Comment:
I would tend to have performEviction take a Vector<std::pair< SecurityOriginData , AccessRecord>> so that we do not need both originRecords and sortedOrigins.
We would then be able to sort more efficiently, since we would not query the map every time we do a comparison.

Reply:
Yes, the thing I am trying to avoid is query the map during sorting.
We should just need to query the item itself